### PR TITLE
tempo tracing interface spec

### PR DESCRIPTION
**Pull request by PietroPasotti**
This PR adds a 'tracing' relation interface, meant to be used to exchange the necessary information for one charm to expose endpoints to which traces can be sent.

### CODE REVIEW

**Review by simskij — CHANGES_REQUESTED** [review]
please add a link from the main readme.

**Comment by gruyaume** [comment]
Can you also please add a description to the PR, this is useful when reviewing.

File: interfaces/tracing/v0/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# `tracing`
+
+## Usage
+
+This relation interface is meant to exchange the necessary information for a charm to integrate with a Tempo-compliant tracing backend.
```
Comment (gruyaume) [inline interfaces/tracing/v0/README.md:5]: It's not just for the charm that integrates with a tempo-compliant tracing backend, it's also for that tempo-compliant tracing backend. I think the generic `This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.`

File: interfaces/tracing/v0/schemas/provider.json [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                        "type": "otlp_grpc",
+                        "port": 1234
+                    },
+                    {
+                        "type": "otlp_http",
+                        "port": 1233
+                    }
+                ]
+            },
+            "related-units": {
```
Comment (gruyaume) [inline interfaces/tracing/v0/schemas/provider.json:23]: In most json schema specs here we don't include the generic data that we'll find in all interfaces like `related_units`

File: interfaces/tracing/v0/schemas/provider.json [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    "$schema": "http://json-schema.org/draft-07/schema",
+    "$id": "https://canonical.github.io/charm-relation-interfaces/interfaces/tracing/schemas/provider.json",
+    "type": "object",
+    "title": "`tracing` provider schema",
+    "description": "The `tracing` root schema comprises the entire provider databag for this interface.",
+    "default": {},
+    "examples": [
+        {
+            "application-data": {
+                "hostname": "tempo.example.org",
```
Comment (gruyaume) [inline interfaces/tracing/v0/schemas/provider.json:11]: Should we make `hostname` and `ingester-ports` required fields?

File: interfaces/tracing/v0/schemas/provider.json [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                "ingester-ports": [
+                    {
+                        "type": "otlp_grpc",
```
Comment (gruyaume) [inline interfaces/tracing/v0/schemas/provider.json:14]: Should we make `type` and `port` required fields?

**Comment by simskij** [comment]
Bump @PietroPasotti

File: interfaces/tracing/v0/README.md (hunk 1 above)
Comment (simskij) [reply interfaces/tracing/v0/README.md:5]: Agreed.

File: interfaces/tracing/v0/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+## Direction
+
+Tracing is a push-based system. The tracing backend, also called an ingester, can support a number of different protocols, such as [otlp-grpc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc) or [otlp-http](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
+The tracing backend **provides**, for each protocol it supports, an endpoint at which the server is ready to accept that protocol. So the directionality of the relation flows from the observer: the trace ingester (a Tempo(-compliant) backend), to the observed: the application producing the traces, such as a mattermost charm.
+
+The requirer is supposed to receive the supported protocols and use it to determine where to send the traces.
```
Comment (simskij) [inline interfaces/tracing/v0/README.md:12]:
```suggestion
```

Already described below in the behavior of the requirer.

File: interfaces/tracing/v0/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+flowchart LR
+    Provider -- TracingBackend --> Requirer
+```
+
+## Behavior
+### Provider
+
+- Is expected to publish the hostname at which the server is reachable.
+- Is expected to run a server supporting one or more tracing protocols such as [OTLP](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification).
+- Is expected to publish via relation data, for each protocol it accepts, the port at which the server is listening.
```
Comment (simskij) [inline interfaces/tracing/v0/README.md:28]:
```suggestion
- Is expected to publish, for each protocol it accepts, the port at which the server is listening.
```

**Review by simskij — COMMENTED** [review]
Looks good in general, some minor nits on language. Once the open conversations have been resolved, I think this PR is ready to merge. 🔥

File: interfaces/tracing/v0/README.md (hunk 1 above)
Comment (PietroPasotti) [inline interfaces/tracing/v0/README.md:5]:
```suggestion
This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.
```

**Review by simskij — APPROVED** [review]

**Comment by PietroPasotti** [comment]
@gruyaume @simskij I reworked quite a bit the interface, we need a new review.

File: interfaces/tracing/v0/README.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
+# `tracing`
+
+## Usage
+
+This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.
+
+## Direction
+
+Tracing is done in a push-based fashion. The receiving endpoint of the tracing backend, also referred to as an ingester, can support a number of different protocols, such as [otlp-grpc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc) and [otlp-http](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
+The tracing backend **provides**, for each protocol it supports, an endpoint at which the server is ready to accept that protocol. So the directionality of the relation flows from the observer: the trace ingester (a Tempo(-compliant) backend), to the observed: the application producing the traces, such as a mattermost charm.
```
Comment (sed-i) [inline interfaces/tracing/v0/README.md:10]: Strictly speaking, a relation does not have directionality.

File: interfaces/tracing/v0/README.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+
+This follows the convention of other push models such as `loki_push_api`.
+
+We call the data structure that is exchanged via this interface 'tracing backend'.
+
+```mermaid
+flowchart LR
+    Provider -- TracingBackend --> Requirer
```
Comment (sed-i) [inline interfaces/tracing/v0/README.md:19]: Seems clearer when we use the domain's terms next to provider/requirer:
```suggestion
    provider[Ingester (provider)] -- TracingBackend --> requirer[Observed app (requirer)]
```

File: interfaces/tracing/v0/README.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+```
+
+## Behavior
+### Provider
+
+- Is expected to publish the hostname at which the server is reachable.
```
Comment (sed-i) [inline interfaces/tracing/v0/README.md:25]: "Hostname" seems to restricting as it excludes scheme, path, ports etc.
"Address" is too ambiguous.
How about URL?

```suggestion
- Is expected to publish the URL at which the server (ingester) is reachable.
```

File: interfaces/tracing/v0/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+- Is expected to run a server supporting one or more tracing protocols such as [OTLP](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification).
+- Is expected to publish, for each protocol it accepts, the port at which the server is listening.
+
+
+### Requirer
+
+- Is expected to push traces to one or more of the supported endpoints.
```
Comment (sed-i) [inline interfaces/tracing/v0/README.md:32]: "push traces" is a workload capacity, not a relation/charm capacity.
Iiuc, in the context of this repo, "requirer" and "provider" do not include the workload.

File: interfaces/tracing/v0/README.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+- zipkin
+
+[\[JSON Schema\]](./schemas/provider.json)
+
+
+#### Example
+```yaml
+# unit_data: <empty> 
+application_data: 
+  hostname: "http://foo.bar/my-model-my-unit-0"
```
Comment (sed-i) [inline interfaces/tracing/v0/README.md:55]: Strictly speaking, "hostname" would be just `foo.bar`.

Need to figure out how to specify [URL parts](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) in relation data.
If the port is different, is there guarantee that the other URL parts remain the same?

File: interfaces/tracing/v0/README.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  ingester-ports: 
+    - type: otlp_grpc
+      port: 1234
+    - type: otlp_http
+      port: 5678
+```
+
+### Requirer
+
+The requirer side is not expected to publish any data via this relation's databags.
```
Comment (sed-i) [inline interfaces/tracing/v0/README.md:65]: Haven't looked into tracing yet, but would we want the requirer to announce its juju topology for labelling of traces?

File: interfaces/tracing/v0/charms.yaml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+providers:
+  - name: tempo-k8s
+    url: https://github.com/PietroPasotti/tempo-k8s-operator
```
Comment (sed-i) [inline interfaces/tracing/v0/charms.yaml:3]:
```suggestion
    url: https://github.com/canonical/tempo-k8s-operator
```

File: interfaces/tracing/v0/README.md (hunk 7 above)
Comment (PietroPasotti) [reply interfaces/tracing/v0/README.md:10]: fair enough

File: interfaces/tracing/v0/README.md (hunk 12 above)
Comment (PietroPasotti) [reply interfaces/tracing/v0/README.md:65]: The thing producing the traces would be responsible for injecting that kind of metadata in the traces. So no relabeling is required.

File: interfaces/tracing/v0/README.md (hunk 10 above)
Comment (PietroPasotti) [reply interfaces/tracing/v0/README.md:32]: in my experience this distinction is kind of blurry in this repository, although it does make sense to me.

File: interfaces/tracing/v0/README.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+## Usage
+
+This relation interface describes the expected behavior of any charm claiming to be able to provide or consume a tempo-compliant tracing server.
+
+## Direction
+
+Tracing is done in a push-based fashion. 
+The receiving endpoint of the tracing backend, also referred to as an ingester, can support a number of different protocols, such as [otlp-grpc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc) and [otlp-http](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
+The tracing backend exposes, for each protocol it supports, an endpoint at which the server is ready to accept that protocol. 
+So the directionality of the relation flows from the observed, the application producing the traces, to the observer (aka the trace ingester, tha is, a Tempo(-compliant) backend).
```
Comment (simskij) [inline interfaces/tracing/v0/README.md:12]:
```suggestion
So the directionality of the relation flows from the observed, the application producing the traces, to the observer (aka the trace ingester, that is, a Tempo(-compliant) backend).
```

**Review by simskij — APPROVED** [review]

**Pull request merged**